vhost-device-spi/src/main.rs

```diff
@@ -0,0 +1,176 @@
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct SpiArgs {
+    /// Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
+    #[clap(short, long)]
+    socket_path: String,
+
+    /// Number of guests (sockets) to connect to.
+    #[clap(short = 'c', long, default_value_t = 1)]
+    socket_count: usize,
+
+    /// SPI device full path
+    #[clap(short = 'l', long)]
+    device: String,
+}
+
+#[derive(PartialEq, Debug)]
+struct SpiConfiguration {
+    socket_path: String,
+    socket_count: usize,
+    device: String,
+}
+
+impl SpiConfiguration {
+    fn try_from(args: SpiArgs) -> Result<Self> {
+        if args.socket_count == 0 {
+            return Err(Error::SocketCountInvalid(0));
+        }
+
+        Ok(SpiConfiguration {
+            socket_path: args.socket_path.trim().to_string(),
+            socket_count: args.socket_count,
+            device: args.device.trim().to_string(),
+        })
+    }
+}
+
+fn start_backend<D: 'static + SpiDevice + Send + Sync>(args: SpiArgs) -> Result<()> {
+    let config = SpiConfiguration::try_from(args)?;
+
+    let spi_dev = D::open(&config.device).map_err(Error::SpiFailure)?;
+
+    let spi_ctrl = Arc::new(SpiController::<D>::new(spi_dev).map_err(Error::SpiFailure)?);
+
+    let mut handles = Vec::new();
+
+    for i in 0..config.socket_count {
+        let socket = config.socket_path.to_owned() + &i.to_string();
+        let spi_ctrl = spi_ctrl.clone();
+
+        let handle: JoinHandle<Result<()>> = spawn(move || loop {
+            // A separate thread is spawned for each socket and can connect to a separate guest.
+            // These are run in an infinite loop to not require the daemon to be restarted once a
+            // guest exits.
+            //
+            // There isn't much value in complicating code here to return an error from the
+            // threads, and so the code uses unwrap() instead. The panic on a thread won't cause
+            // trouble to other threads/guests or the main() function and should be safe for the
+            // daemon.
+            let backend = Arc::new(RwLock::new(
+                VhostUserSpiBackend::new(spi_ctrl.clone()).map_err(Error::CouldNotCreateBackend)?,
+            ));
+
+            let mut daemon = VhostUserDaemon::new(
+                String::from("vhost-device-spi-backend"),
+                backend.clone(),
+                GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+            )
+            .map_err(Error::CouldNotCreateDaemon)?;
+
+            daemon.serve(&socket).map_err(Error::ServeFailed)?;
+        });
+
+        handles.push(handle);
+    }
+
+    for handle in handles {
+        handle.join().map_err(std::panic::resume_unwind).unwrap()?;
+    }
```

Comment:
This waits for each handle sequentially and will not join handles that are next. In other crates this was changed to have each thread send a signal to the parent process:

https://github.com/rust-vmm/vhost-device/blob/bd295df4db6b60649b5a26032c5f7f2a1620f8c5/vhost-device-input/src/main.rs#L151-L159

Reply:
Okay, will update.